// next/dist/experimental/testmode/playwright/next-fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "applyNextFixture", {
    enumerable: true,
    get: function() {
        return applyNextFixture;
    }
});
const _pageroute = require("./page-route");
class NextFixtureImpl {
    constructor(testId, worker, page){
        this.testId = testId;
        this.worker = worker;
        this.page = page;
        this.fetchHandler = null;
        this.page.route("**", (route)=>(0, _pageroute.handleRoute)(route, page, this.fetchHandler));
    }
    teardown() {
        this.worker.cleanupTest(this.testId);
    }
    onFetch(handler) {
        this.fetchHandler = handler;
        this.worker.onFetch(this.testId, handler);
    }
}
async function applyNextFixture(use, { testInfo , nextWorker , page , extraHTTPHeaders  }) {
    const fixture = new NextFixtureImpl(testInfo.testId, nextWorker, page);
    page.setExtraHTTPHeaders({
        ...extraHTTPHeaders,
        "Next-Test-Proxy-Port": String(nextWorker.proxyPort),
        "Next-Test-Data": fixture.testId
    });
    await use(fixture);
    fixture.teardown();
}
